rust/main/chains/hyperlane-ton/src/contracts/mailbox.rs

```diff
@@ -0,0 +1,672 @@
+            .seqno as u32;
+
+        let message = self
+            .signer
+            .wallet
+            .create_external_message(
+                now + 60,
+                seqno,
+                vec![ArcCell::new(transfer_message.clone())],
+                false,
+            )
+            .map_err(|e| {
+                ChainCommunicationError::CustomError(format!(
+                    "Failed to create external message: {:?}",
+                    e
+                ))
+            })?;
+
+        let boc = BagOfCells::from_root(message.clone())
+            .serialize(true)
+            .map_err(|e| {
+                ChainCommunicationError::CustomError(format!("Failed to serialize BOC: {:?}", e))
+            })?;
+
+        let boc_str = general_purpose::STANDARD.encode(&boc);
+        info!("create_external_message:{:?}", boc_str);
+
+        let tx = self.provider.send_message(boc_str).await.map_err(|e| {
+            ChainCommunicationError::CustomError(format!(
+                "Failed to send message in provider{:?}",
+                e
+            ))
+        })?;
+
+        info!("Tx hash:{:?}", tx.message_hash);
+
+        self.provider.wait_for_transaction(tx.message_hash).await
+    }
+
+    async fn process_estimate_costs(
+        &self,
+        _message: &HyperlaneMessage,
+        _metadata: &[u8],
+    ) -> ChainResult<TxCostEstimate> {
+        Ok(TxCostEstimate {
+            gas_limit: U256::zero(),
+            gas_price: FixedPointNumber::zero(),
+            l2_gas_limit: None,
+        })
+    }
+
+    fn process_calldata(&self, _message: &HyperlaneMessage, _metadata: &[u8]) -> Vec<u8> {
+        todo!()
+    }
+}
+
+#[derive(Debug)]
+pub struct TonMailboxIndexer {
+    pub mailbox: TonMailbox,
+}
+
+#[async_trait]
+impl Indexer<HyperlaneMessage> for TonMailboxIndexer {
+    async fn fetch_logs_in_range(
+        &self,
+        range: RangeInclusive<u32>,
+    ) -> ChainResult<Vec<(Indexed<HyperlaneMessage>, LogMeta)>> {
+        let start_block = *range.start();
+        let end_block = *range.end();
+
+        let fetch_block_info = |block: u32| async move {
+            self.mailbox
+                .provider
+                .get_blocks(
+                    -1,          // masterchain (workchain = -1)
+                    None,        // shard
+                    None,        // Block block seqno
+                    Some(block), // Masterchain block seqno
+                    None,
+                    None,
+                    None,
+                    None,
+                    None,
+                    None,
+                    None,
+                )
+                .await
+                .map_err(|e| {
+                    ChainCommunicationError::CustomError(format!(
+                        "Failed to fetch block info for block {}: {:?}",
+                        block, e
+                    ))
+                })?
+                .blocks
```

Comment:
I suppose that it will be useful to extract this call into a separate method and use it here as well as in other `Indexer`s